pkg/multimap/composite.go

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2019-Present Pivotal Software, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package multimap
+
+import "fmt"
+
+// Composite is an interface to a consistent collection of mappings from string to string. The mappings are consistent
+// in the sense that no two mappings map a given value to distinct results.
+type Composite interface {
+	// Add updates the composite mapping by adding a mapping with the given name. If a mapping with the given name
+	// already exists, it is replaced providing it is consistent with all other mappings. If the named mapping is not
+	// consistent with all other mappings, the mapping is removed from the composite mapping and error is returned.
```

Comment:
does this mean I can't remove an image from an image map?

Reply:
how are collisions managed?

Reply:
> does this mean I can't remove an image from an image map?

No - removing an image would never introduce an inconsistency. An inconsistency is where a single original image maps to two or more distinct relocated images.

Reply:
> how are collisions managed?

When you try to add a mapping which would cause a collision, the whole add operation fails and the mapping is removed from the composite state. (The other option when replacing a mapping would be to leave the old mapping behind, but this seemed worse.)

Reply:
Continued in https://github.com/pivotal/kubernetes-image-mapper/pull/2